### PR TITLE
Add a new weariness panel display

### LINKED ISSUE #44428: A weariness indicator should be in the sidebar
# Is your feature request related to a problem? Please describe.
Since weariness can drastically multiply the amount of time certain activities take, I think it should be something the player is able to be aware of at a glance, so they can monitor it without having to interrupt their 'cool down' activity.

# Describe the solution you'd like
Add some indication of weariness to the sidebar. With the current, build 11035, implementation of weariness, I have to regularly interrupt my reading/waiting to check if my level of weariness has gone down. Log messages are helpful, but can be hard to keep track of when the ~~goddamn ducks keep flying over the flowing deep water~~ log is filled with other messages, and it's not always straightforward to translate a lack of log messages to mean that I am no longer weary.

# Describe alternatives you've considered
Keep interrupting my waiting to check the character sheet.

# Additional context
None.

### PR DESCRIPTION
**Pull request by AlexMooney**
#### Summary

SUMMARY: Interface "Add a new panel display for weariness"

#### Purpose of change

Weariness changes spam the message log and are in the character sheet, though they should be visible on the side panel.

Fixes #44428

#### Describe the solution

1) Add a weariness bar that goes down as you get tired.
2) Add a display of the *six* different penalties that you could suffer based on your current weariness, and highlight the category that corresponds to your recent activity level.

image

#### Describe alternatives you've considered

- ~Show slowdown factor(s) instead of a HP bar.~ Added!
- Maybe change the "Speed" label for the crazy row?  It should be 5 characters to fit "narrow" panel,
- Maybe change the left-most "100" to "Rest", since it would never be penalized?

#### Testing

Did some deconstruction of furniture to drain weariness and watched the bar go down and back up with rest.

#### Additional context

I cannot display how bad the activity the player is doing is draining weariness in real time; only the worst activity in the last ~5 minute chunk is tracked.  Hopefully it will be possible after the new and improved weariness tracking over at https://github.com/CleverRaven/Cataclysm-DDA/issues/44428.

### CODE REVIEW

**Comment by Raikiri** [comment]
Can we also please add information somewhere on what exact effects weariness has? Like, I noticed that "extremely weary" drastically nerfs my attack speed, but where to I find exactly how much it tanks? What other effects does it have? I really wish that stuff was in the UI.

**Comment by anothersimulacrum** [comment]
The effect weariness has on activities is entirely dependent on the activity level of those activities, so activity level must be communicated to communicate effects.

**Comment by AlexMooney** [comment]
> The effect weariness has on activities is entirely dependent on the activity level of those activities, so activity level must be communicated to communicate effects.

It sounds like the weariness speed debuffs should be consolidated to just 1 or 2 penalties instead of a per-activity penalty, so that we could make a more generic but correct display to the player.

**Comment by anothersimulacrum** [comment]
> > The effect weariness has on activities is entirely dependent on the activity level of those activities, so activity level must be communicated to communicate effects.
> 
> It sounds like the weariness speed debuffs should be consolidated to just 1 or 2 penalties instead of a per-activity penalty, so that we could make a more generic but correct display to the player.

It would require substantial changes in how the entire system operates and the whole functioning of it, so no.

File: src/panels.cpp
Comment (anothersimulacrum) [inline src/panels.cpp:2072]: 1. Less auto, please.
2. These HP bars are very unclear. The exertion bar appears to do nothing, whether I'm digging a pit, or just sitting there waiting. A HP-bar style display would work for how far you are to the next level of weariness, but that's going to be fairly closely tied to how weariness is calculated, not how this is currently implemented. Exertion and weariness should be state displays, like hunger, or tiredness, or temperature.

File: src/suffer.cpp
Comment (anothersimulacrum) [inline src/suffer.cpp:1404]: 3. This is really spammy, I don't think there's much reason for it to be so much more often.
4. The display at the end is the floating point numbers that these activities correspond to, which means nothing. These should be either names, or just be removed. And I'm not sure this is of much value to display in addition to the panel.

**Review by anothersimulacrum — CHANGES_REQUESTED** [review]

File: src/suffer.cpp
Comment (AlexMooney) [reply src/suffer.cpp]: This was a git mistake from my trying to figure out a way to get the exertion bar to load.  Didn't realize I had committed it.

File: src/panels.cpp
Comment (AlexMooney) [reply src/panels.cpp:2072]: 1. Sure, though auto is used almost everywhere else for `get_hp_bar`, probably because `std::pair<std::string, nc_color>` is a mess.
2. As mentioned in the original post, the exertion bar does nothing.  It's intended to reflect how badly your most recent activity is draining your weariness, but I'll just remove it.

**Comment by AlexMooney** [comment]
I've added a speed percentage readout too; it's only accurate for `EXTRA_EXERCISE`, which is the worst penalty.  I've labeled it "Speed", which I don't think we should use since that conflicts with the existing Speed that tells you how many moves the previous action consumed.

**Comment by Raikiri** [comment]
> The effect weariness has on activities is entirely dependent on the activity level of those activities, so activity level must be communicated to communicate effects.

i think it's worth at least communicating best to worst case, something like "actions take from 25% to 500% more time depending on their activity level"

**Comment by AlexMooney** [comment]
@Raikiri the speed penalty is part of the bar now (though the screenshot has a colored percentage and I've removed the color from that part)

**Comment by AlexMooney** [comment]
I've added a second line so that we can show all of the speed penalties.  The activity level that you're considered to be at for weariness updates and metabolism is highlighted.

Resting:
image
Walking:
image
Deconstructing:
image

**Comment by anothersimulacrum** [comment]
I've made a PR demonstrating an alternate mode of display at #46340.

**Comment by Raikiri** [comment]
I actually prefer AlexMooney's version because it immediately shows how much penalties I would get on activities that I'm not doing right now. And having 4 display points for one game mechanic in anothersimulacrum seems like it's splitting attention a bit too much.

**Comment by AlexMooney** [comment]
I think we should combine the second line of my PR with the transitions from @anothersimulacrum's PR.

Something sort of like this:
```
Weary: Fresh         |||\.  <-- level and progress till next level
Malus: 100/100/100/100/100  <-- All of the penalties with the highlighting
```

I think we should also merge @anothersimulacrum's changes to the weariness status messages separately from the side panel, because they're clearly an improvement.

**Comment by anothersimulacrum** [comment]
> We should also merge @anothersimulacrum's changes to the weariness status messages separately from the side panel, because they're clearly an improvement, but are only tangentially related to adding a panel display.

They're only related in that I'm going to insist that we have an activity level display in the sidebar before it's merged :)

Past that, I think a final display of something like this (labels aren't necessary except to show you what it is for discussion) would be a good thing, what do you think about it?
```
Weary: Fresh     Transition: |||\.
Activity: Moderate 100/100/100/100
```
Either in a single weariness panel or two separate ones (but that's a bit over the top, IMO).
It would probably be useful to add a 'where you're headed' indicator in addition to the transition bar, but that can be an exercise for another time.

The final issue is that the display in the effects menu and display here inconsistent, so the malus bar could be changed to percentages (which makes formatting annoying with extreme level), like this: `100%/100%/100%/133%/167%` (extreme can go to 1000%). Or it could be changed to multipliers, which also could get a little messy with formatting, but in general would look something like: `1x/1x/1x/1.33x/1.67x`.

Then as a final, and totally unrelated to any of this panel stuff except that it's adding info about this system, thing, a `melee combat cost multiplier` could/should be added to the speed section of the `@` menu (separate from the effects).

If this sounds good to you and you want to go ahead, you can grab whatever you want/need of my PR and work on that here, and I can close my PR, or however you want to go ahead here.

**Comment by AlexMooney** [comment]
> > We should also merge @anothersimulacrum's changes to the weariness status messages separately from the side panel, because they're clearly an improvement, but are only tangentially related to adding a panel display.
> 
> They're only related in that I'm going to insist that we have an activity level display in the sidebar before it's merged :)

Why _wouldn't_ we improve the status messages right away while we keep working on the panel bar? :confused: 

> Past that, I think a final display of something like this (labels aren't necessary except to show you what it is for discussion) would be a good thing, what do you think about it?
> 
> ```
> Weary: Fresh         Transition: |||\.
> Activity: Moderate 100/100/100/100/100
> ```
> The final issue is that the display in the effects menu and display here inconsistent, so the malus bar could be changed to percentages (which makes formatting annoying with extreme level), like this: `100%/100%/100%/133%/167%` (extreme can go to 1000%). Or it could be changed to multipliers, which also could get a little messy with formatting, but in general would look something like: `1x/1x/1x/1.33x/1.67x`.

In the Labels Narrow layout we have enough characters for `Malus: 100/200/300/500/900/1000` but not a single character more (and that last character is not used by the rest of the status bar), so we won't be able to have the activity level spelled out, nor will we have room for the `%` after each number.

```
L ARM: |||||      R LEG: ||||.
Weary: Xtrem Transition: |||\.    <- Only 5 characters per description if we include "Transition"
Malus: 100/200/300/500/900/1000   <- usually aligned with the end of the transition bar, but 1000 will bump it out
```

I'd prefer to drop the big label so that we don't have to abbreviate the descriptions:
```
L ARM: |||||      R LEG: ||||.
Weary: Moderate          |||\.    <- I think people can read this without a second label
Malus: 100/100/100/125/250/500
```

I think percentages that have implied "%" is the only way to make it work if we want to use the reciprocals.  Could also change the menu to say "10% speed" instead of "+900% duration" or whatever, but I'm fine making it 1000 instead of 10.
 
> Then as a final, and totally unrelated to any of this panel stuff except that it's adding info about this system, thing, a `melee combat cost multiplier` could/should be added to the speed section of the `@` menu (separate from the effects).

I agree; it should also be added in another PR, and it would definitely be helpful.

> If this sounds good to you and you want to go ahead, you can grab whatever you want/need of my PR and work on that here, and I can close my PR, or however you want to go ahead here.

Not sure yet.  I think your PR is closer to what we want.  Your instantaneous activity levels is something we want, as are the description changes, and the level labels and the bar being tied to the transition.  If you want to take it, I'm fine closing this; if you're sick of side panel bars, I can merge some of your commits in here.

**Comment by anothersimulacrum** [comment]
> > > We should also merge @anothersimulacrum's changes to the weariness status messages separately from the side panel, because they're clearly an improvement, but are only tangentially related to adding a panel display.
> > 
> > 
> > They're only related in that I'm going to insist that we have an activity level display in the sidebar before it's merged :)
> 
> Why _wouldn't_ we improve the status messages right away while we keep working on the panel bar? confused

Because it's not really an improvement to add information that almost nobody knows what to do with.
Sure, now we know information about the effects, but we have no clue when it applies, what these 'activity levels' are, or when what actions are these levels.

We _must_ make this information accessible before we add this to the tooltips, or it is effectively meaningless.

> > Past that, I think a final display of something like this (labels aren't necessary except to show you what it is for discussion) would be a good thing, what do you think about it?
> > ```
> > Weary: Fresh         Transition: |||\.
> > Activity: Moderate 100/100/100/100/100
> > ```
> > 
> > 
> > The final issue is that the display in the effects menu and display here inconsistent, so the malus bar could be changed to percentages (which makes formatting annoying with extreme level), like this: `100%/100%/100%/133%/167%` (extreme can go to 1000%). Or it could be changed to multipliers, which also could get a little messy with formatting, but in general would look something like: `1x/1x/1x/1.33x/1.67x`.
> 
> In the Labels Narrow layout we have enough characters for `Malus: 100/200/300/500/900/1000` but not a single character more (and that last character is not used by the rest of the status bar), so we won't be able to have the activity level spelled out, nor will we have room for the `%` after each number.
> 
> ```
> L ARM: |||||      R LEG: ||||.
> Weary: Xtrem Transition: |||\.    <- Only 5 characters per description if we include "Transition"
> Malus: 100/200/300/500/900/1000   <- usually aligned with the end of the transition bar, but 1000 will bump it out
> ```
> 
> I'd prefer to drop the big label so that we don't have to abbreviate the descriptions:
> 
> ```
> L ARM: |||||      R LEG: ||||.
> Weary: Moderate          |||\.    <- I think people can read this without a second label
> Malus: 100/100/100/125/250/500
> ```

Totally agree, the label was just for illustrating what it was to anyone reading my post, because it wasn't necessarily super clear (though I guess it was the only thing it could have been in context).
 
> I think percentages that have implied "%" is the only way to make it work if we want to use the reciprocals. Could also change the menu to say "10% speed" instead of "+900% duration" or whatever, but I'm fine making it 1000 instead of 10.

I think we either need to add another line containing this/the activity level for sidebars too narrow to have both the activity level and this, or just drop using the reciprocals entirely, because I consider an absolute necessity to have the activity level display.

> > If this sounds good to you and you want to go ahead, you can grab whatever you want/need of my PR and work on that here, and I can close my PR, or however you want to go ahead here.
> 
> Not sure yet. I think your PR is closer to what we want. Your instantaneous activity levels is something we want, as are the description changes, and the level labels and the bar being tied to the transition. If you want to take it, I'm fine closing this; if you're sick of side panel bars, I can merge some of your commits in here.

I mostly wrote my PR because I wanted to show how I thought it should look without dictating to you to follow my arbitrary whims, so I'm totally happy to just stop working on this and leave it up to you.

**Comment by Raikiri** [comment]
Maybe instead of listing maluses for all possible activities like this:
`Malus: 100/200/300/500/900/1000`
Just list them like this instead
`Malus: x6.0/x10.0` where left one is the malus for the last activity performed and the right one is the maximum possible malus for current weariness level?

**Comment by anothersimulacrum** [comment]
That's a good idea, but still doesn't fix it for the worst case, which is what I think we should plan for.

**Comment by AlexMooney** [comment]
Finally got around to pulling in the labels and inverting the penalties on the bar.  I think this is good to go, and someone else can pretty it up some more.

image

**Review by anothersimulacrum — CHANGES_REQUESTED** [review]
This needs to either drop de19050d792ec9a81e5633818f4d1b745a5e6bcb, or display activity level in sidebar normally.
Those _cannot_ be merged without showing activity level in the sidebar, because they are giving information that is reliant on information that is entirely hidden to the player.

**Comment by AlexMooney** [comment]
The `Speed` line shows the player which activity level they're operating at.

**Comment by anothersimulacrum** [comment]
Not really. The words shown in the descriptions _must_ be shown in the UI in the context of activity. My whole issue is that the words used there are meaningless, because they aren't displayed anywhere. They aren't even the words used internally for activity level.

**Comment by moxian** [comment]
Drive-by comment: As someone who knows weariness exists and that when you get weary, it says something along the lines of "you are now weary and thus slower" (forget the exact wording), i have NO idea how to interpret `100/100/100/125/250/500` display.

IF (and this is assumption I've just made on the spot, with the helpful context in this thread) that this represents how much slower different intensity of action becomes (first for easy things, second for harder, last one for digging rocks), it really should not have all the numbers on the status bar all the time.
Instead I would much rather have:
```
Weary: ||...
[current] Excercise [level]: Light
```
and move the numbers to the description in @ panel, i.e.:
20210119-020250-cataclysm-tiles

**Comment by AlexMooney** [comment]
Go ahead with your PR then @anothersimulacrum.  :shrug:

**Pull request closed without merging**
